botnim/document_parser/dynamic_extractions/pdf_extraction/pdf_pipeline.py

```diff
@@ -0,0 +1,461 @@
+                logger.warning(f"No PDF files were successfully processed for source '{source_name}'")
+                return True
+            
+            logger.info(f"Successfully processed {len(all_results)} entities from {len(pdf_files)} PDF files for source '{source_name}'")
+            
+            # Generate CSV output
+            fieldnames = [f.name for f in source_config.fields]
+            flat_data = []
+            for result in all_results:
+                row = flatten_for_csv(result, fieldnames)
+                flat_data.append(row)
+            
+            # Write CSV file (for debugging/backup)
+            csv_path = write_csv(flat_data, fieldnames, source_name, str(self.output_dir))
+            logger.info(f"CSV output written to: {csv_path}")
+            
+            # Note: Google Sheets upload is done in real-time during processing
+            
+            return True
+            
+        except Exception as e:
+            logger.error(f"Failed to process source '{source_name}': {e}")
+            return False
+    
+    def process_all_sources(self, upload_to_sheets: bool = False,
+                           sheets_credentials: Optional[str] = None,
+                           spreadsheet_id: Optional[str] = None,
+                           replace_sheet: bool = False) -> bool:
+        """
+        Process all sources defined in the configuration.
+        
+        Args:
+            upload_to_sheets: Whether to upload results to Google Sheets
+            sheets_credentials: Path to Google Sheets credentials
+            spreadsheet_id: Google Sheets spreadsheet ID
+            replace_sheet: Whether to replace existing sheet content
+            
+        Returns:
+            True if all sources processed successfully, False otherwise
+        """
+        success = True
+        for source in self.config.sources:
+            if not self.process_source(
+                source.name, upload_to_sheets, sheets_credentials, 
+                spreadsheet_id, replace_sheet
+            ):
+                success = False
+        
+        return success
+
+    def save_metrics(self, output_path: Optional[str] = None):
+        """Save performance metrics to a JSON file."""
+        if self.metrics:
+            metrics_path = output_path or str(self.output_dir / "pipeline_metrics.json")
+            self.metrics.save_metrics(metrics_path)
+            logger.info(f"Performance metrics saved to: {metrics_path}")
+    
+    def print_performance_summary(self):
+        """Print a human-readable performance summary."""
+        if self.metrics:
+            self.metrics.print_summary()
+        else:
+            logger.info("Metrics collection is disabled")
+    
+    def get_performance_summary(self) -> Optional[Dict]:
+        """Get performance summary as a dictionary."""
+        if self.metrics:
+            return self.metrics.get_pipeline_summary()
+        return None
+
+def main():
```

Comment:
let's remove all ad-hoc argument parsing methods and run everything through the cli

Reply:
(relevant to all source files)